plutus-core/exe/Main.hs

```diff
@@ -39,33 +40,53 @@ import           System.CPUTime                                             (get
 import           System.Exit
 import           Text.Printf
 
-
-{- Note [Annotation types] This program now reads and writes
-   CBOR-serialised PLC ASTs.  In all cases we require the annotation
-   type to be ().  There are two reasons for this.  Firstly, ASTs
-   serialised for transmission to the chain will always have unit
-   annotations because we can save space by omitting annotations in
-   the CBOR (using the OmitUnitAnnotations class from CBOR.hs), so it
-   makes sense for the program to deal with these.  Secondly, the
-   annotation type has to be specified when we're deserialising CBOR
-   (in order to check that the AST has the correct type), so it's
-   difficult to deal with CBOR with arbitrary annotation types: fixing
-   the annotation type to be () is the simplest thing to do and fits
-   our use case.
+{- Note [Annotation types] This program now reads and writes CBOR-serialised PLC
+   ASTs.  In all cases we require the annotation type to be ().  There are two
+   reasons for this.  Firstly, ASTs serialised for transmission to the chain
+   will always have unit annotations because we can save space by omitting
+   annotations in the CBOR (using the OmitUnitAnnotations class from CBOR.hs),
+   so it makes sense for the program to deal with these.  Secondly, the
+   annotation type has to be specified when we're deserialising CBOR (in order
+   to check that the AST has the correct type), so it's difficult to deal with
+   CBOR with arbitrary annotation types: fixing the annotation type to be () is
+   the simplest thing to do and fits our use case.
  -}
 
 
-type PlainProgram   = PLC.Program PLC.TyName PLC.Name PLC.DefaultUni ()
-type ParsedProgram  = PLC.Program PLC.TyName PLC.Name PLC.DefaultUni PLC.AlexPosn
+type TypedProgram a = PLC.Program PLC.TyName PLC.Name PLC.DefaultUni a
+type UntypedProgram a = UPLC.Program PLC.Name PLC.DefaultUni a
+
+data Program a =
+      TypedProgram (TypedProgram a)
+    | UntypedProgram (UntypedProgram a)
+    deriving (Functor)
+
+instance (PLC.PrettyBy PLC.PrettyConfigPlc (Program a)) where
+    prettyBy cfg (TypedProgram p)   = PLC.prettyBy cfg p
+    prettyBy cfg (UntypedProgram p) = PLC.prettyBy cfg p
+
 type PlcParserError = PLC.Error PLC.DefaultUni PLC.AlexPosn
 
 
 ---------------- Option parsers ----------------
 
+data Typing = Typed | Untyped
+
+typed :: Parser Typing
+typed = flag Untyped Typed (long "typed" <> short 't' <> help "Use typed Plutus Core")
+
+untyped :: Parser Typing
+untyped = flag Untyped Untyped (long "untyped" <> short 'u' <> help "Use untyped Plutus Core (default)")
+-- ^ NB: default is always Untyped
+
+typingMode :: Parser Typing
+typingMode = typed <|> untyped
+
 data Input = FileInput FilePath | StdInput
 
+-- | Parser for an input stream. If none is specified, default to stdin: this makes use in pipelines easier
 input :: Parser Input
-input = fileInput <|> stdInput
+input = fileInput <|> stdInput <|> pure StdInput
```

Comment:
So did I, but apparently not in this case.  It has "modfiers" which you can use to attach defaults, help text and so on to parser _builders_ (https://hackage.haskell.org/package/optparse-applicative-0.14.2.0/docs/Options-Applicative.html#t:Mod).  If you have something like a `-o <filename>` option this allows you to supply a default for the filename, like `output.txt` or something.  However here we're composing two _parsers_ and I don't think there's any way to say which of the alternatives to use by default, except what I've done here.  I could be missing something though.

Reply:
Sure, but you could make the default for the input/output options be `stdin`, right? You'd have to do it for each option that uses this parser, but that's okay, I think? That way it shows up in the help text too.

Reply:
I don't quite understand that.  For example the input stream really is either `--stdin` or `-i <filename>` which really is an alternative between two separate options.  I think you have to build that using `<|>` and there's nowhere to attach a `Mod` there.

I have just added `(default)` to the `stdInput` and `stdOutput` flags.  That's fragile because you have to change the help text if you change the `input` parser which actually implements the choice, but I don't see a better way to do it.

Reply:
I see what you're saying now. That does seem awkward. Not sure what to do, what's there seems fine.